ethers-solc/src/artifacts.rs

```diff
@@ -683,19 +689,32 @@ impl OutputContracts {
 /// Helper type to implement display for solc errors
 #[derive(Clone, Debug)]
 pub struct OutputDiagnostics<'a> {
-    errors: &'a [Error],
+    compiler_output: &'a CompilerOutput,
     ignored_error_codes: &'a [u64],
 }
 
 impl<'a> OutputDiagnostics<'a> {
     /// Returns true if there is at least one error of high severity
     pub fn has_error(&self) -> bool {
-        self.errors.iter().any(|err| err.severity.is_error())
+        self.compiler_output.has_error()
     }
 
     /// Returns true if there is at least one warning
     pub fn has_warning(&self) -> bool {
-        self.errors.iter().any(|err| err.severity.is_warning())
+        self.compiler_output.has_warning(&self.ignored_error_codes)
+    }
+
+    fn is_test<T: AsRef<str>>(&self, contract_path: T) -> bool {
```

Comment:
Not sure how robust this is and I couldn't really figure out a good way to do it for deps, should the test be "if it sits in node_modules or lib then it is a dependency"?

Reply:
hmm, one idea I have is to leave this completely up to the user,

rn we're using it like this:

https://github.com/gakonst/ethers-rs/blob/371afb9dff2041e8ec8d6d808ce5bf18699b04da/ethers-solc/src/lib.rs#L864

perhaps we can add something like 

```rust
 filter: dyn ContractFilter

trait ContractFilter {
   fn matches(&self, contract: &Contract) -> bool;
}

impl ContractFilter for () {
  fn matches(&self) -> boo {true}
}
```

then implement this for `Fn(&str) -> bool`

so we can add a `CompilerOutput::diagnostics_with_filter(error_codes, dyn ContractFilter)`

Reply:
we could also do that for filtering of warnings in general, so instead of a contract filter etc. we just give the user

- `has_warnings`
- `has_errors`
- a way to iterate the errors and filter them themselves

unless there is a specific reason we have this in ethers-solc as opposed to foundry?

Reply:
sg, so we should map each error to its contract and ignore an error if:

* ignored error code
* contract should be ignored